### PR TITLE
[CALCITE-6015] AssertionError during optimization of EXTRACT expression

### PR DESCRIPTION
**Pull request by mihaibudiu**
While attempting to fix the bug described in the issue I have discovered that many cases for EXTRACT are actually not implemented. So I implemented most of them, and I have improved the validation as well; the case indicated in the JIRA issue will now be rejected statically.

There is one case for EXTRACT which I haven't implemented: `EXTRACT(EPOCH FROM INTERVAL '...' YEAR TO MONTH)` because I wasn't sure how to properly use the convertlets to do it. Help would be appreciated.

This PR also makes some unexpected tiny changes in SqlValidatorImpl and SqlCall, so I would appreciate comments from people who are more knowledgeable.

### CODE REVIEW

**Comment by mihaibudiu** [comment]
This PR implements several instances of the EXTRACT function which weren't supported. I would appreciate a review.

**Comment by mihaibudiu** [comment]
Can anyone please review this PR?

File: core/src/main/java/org/apache/calcite/sql/fun/SqlExtractFunction.java
Comment (snuyanzin) [inline core/src/main/java/org/apache/calcite/sql/fun/SqlExtractFunction.java:122]: should we extract the result of the builder here and others into constants (may be sets instead of lists) to avoid building it each time we come here?

File: core/src/main/java/org/apache/calcite/sql/SqlCall.java
Comment (rubenada) [inline core/src/main/java/org/apache/calcite/sql/SqlCall.java:194]: This change on method visibility, although minor, should be considered a breaking change and IMO mentioned briefly on the RN: any downstream project that had overridden it as `protected` would have a compilation issue with the new version, and would need to adjust their code to override it as `public`.

Comment (mihaibudiu) [reply core/src/main/java/org/apache/calcite/sql/SqlCall.java:194]: what should I do about it?

Comment (rubenada) [reply core/src/main/java/org/apache/calcite/sql/SqlCall.java:194]: Usually in these cases, to make sure the Release Manager won't forget about it, what we do is editing the `history.md`: in that file, at the top, there is the "skeleton" with the sections of the next release (commented out), you could add a point in there in the "breaking changes" section mentioning this.

Comment (mihaibudiu) [reply core/src/main/java/org/apache/calcite/sql/SqlCall.java:194]: I have added this to history.md. Thank you for pointing this out.

File: core/src/main/java/org/apache/calcite/sql/fun/SqlExtractFunction.java
Comment (mihaibudiu) [reply core/src/main/java/org/apache/calcite/sql/fun/SqlExtractFunction.java]: I am now using static fields for these constant lists.

Comment (rubenada) [reply core/src/main/java/org/apache/calcite/sql/fun/SqlExtractFunction.java]: Why not using sets as @snuyanzin suggested?

File: site/_docs/history.md
Comment (rubenada) [inline site/_docs/history.md:60]: Thanks for including this. I'd suggest to give a bit more info, e.g. something a long the lines:

In the context of  [<a href="https://issues.apache.org/jira/browse/CALCITE-6015">CALCITE-6015</a>] the visibility of the method `SqlCall.getCallSignature` has been converted from `protected` to `public`, so any subclass overriding it might need to be adjusted accordingly.

Comment (mihaibudiu) [reply site/_docs/history.md]: @rubenada changed List to Set and updated the history file.

**Comment by rubenada** [comment]
Thanks @mihaibudiu , no further comment from my side. LGTM.

**Comment by mihaibudiu** [comment]
Squashed the commits in preparation for merging.

File: site/_docs/history.md
Comment (rubenada) [inline site/_docs/history.md:60]: @mihaibudiu I just noticed, this comment should better be placed a few lines up, right below  the 
```
#### Breaking Changes
{: #breaking-1-37-0}
```

Comment (mihaibudiu) [reply site/_docs/history.md]: sorry about that, I have fixed this, and also squashed the commits

**Comment by rubenada** [comment]
Thanks @mihaibudiu !
@snuyanzin  do you have any further remark? Otherwise IMO this PR is in a good shape to be merged.

**Review by snuyanzin — APPROVED** [review]
Thanks for asking 
and thanks for addressing comments
it looks ok from my side

File: testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java
Comment (mihaibudiu) [inline testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java:10849]: noticed some duplicated tests, I will delete these before merging.

**Pull request merged**
